android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/schedulers/CalendarScheduler.java

```diff
@@ -0,0 +1,223 @@
+import com.cronutils.model.definition.CronDefinition;
+import com.cronutils.model.definition.CronDefinitionBuilder;
+import com.cronutils.model.time.ExecutionTime;
+import com.cronutils.parser.CronParser;
+import com.raizlabs.android.dbflow.annotation.Column;
+import com.raizlabs.android.dbflow.annotation.PrimaryKey;
+import com.raizlabs.android.dbflow.annotation.Table;
+import com.raizlabs.android.dbflow.structure.BaseModel;
+
+import org.joda.time.DateTime;
+import org.json.JSONException;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+
+import host.exp.exponent.notifications.ExponentNotificationManager;
+import versioned.host.exp.exponent.modules.api.notifications.SchedulersManagerProxy;
+import versioned.host.exp.exponent.modules.api.notifications.interfaces.SchedulerInterface;
+import versioned.host.exp.exponent.modules.api.notifications.SchedulersDatabase;
+
+import static com.cronutils.model.CronType.QUARTZ;
+
+@Table(databaseName = SchedulersDatabase.NAME)
+public class CalendarScheduler extends BaseModel implements SchedulerInterface {
+
+  private ArrayList<String> mTriggeringActions = (ArrayList<String>) Arrays.asList(null,
+      Intent.ACTION_BOOT_COMPLETED,
+      Intent.ACTION_TIME_CHANGED,
+      Intent.ACTION_TIMEZONE_CHANGED);
+
+  private Context mApplicationContext;
+
+  public HashMap<String, Object> details;
+
+  // -- model fields --
+
+  @Column
+  @PrimaryKey
+  int id;
+
+  @Column
+  int notificationId;
+
+  @Column
+  String experienceId;
+
+  @Column
+  boolean repeat;
+
+  @Column
+  boolean scheduled = false;
+
+  @Column
+  String serializedDetails;
+
+  @Column
+  String calendarData;
+
+  // -- scheduler methods --
+
+  @Override
+  public void setApplicationContext(Context context) {
+    mApplicationContext = context.getApplicationContext();
+  }
+
+  @Override
+  public boolean schedule(String action) {
+    if (!mTriggeringActions.contains(action)) {
+      return true;
+    }
+    long nextAppearanceTime = 0;
+    
+    try {
+      nextAppearanceTime = getNextAppearanceTime();
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+    
+    ensureDetails();
+    try {
+      getManager().schedule(experienceId, notificationId, details, nextAppearanceTime, null);
+    } catch (ClassNotFoundException e) {
+      e.printStackTrace();
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public String getIdAsString() {
+    return Integer.valueOf(id).toString();
+  }
+
+  private void ensureDetails() {
+    try {
+      details = HashMapSerializer.deserialize(serializedDetails);
+    } catch (JSONException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void cancel() {
+      getManager().cancel(experienceId, notificationId);
+  }
+
+  @Override
+  public boolean canBeRescheduled() {
+    return repeat || (!scheduled);
+  }
+
+  @Override
+  public void scheduled() {
```

Comment:
Not the best name for this method. 😕 Maybe `setScheduled()` or `onPostSchedule()`? I'm not sure